File: leetcode/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans, l, s_set = 0, 0, set()
        for v in s:
            while (v in s_set):
                s_set.remove(s[l])
                l += 1
            s_set.add(v)
            ans = max(ans, len(s_set))
        return ans
        
# @lc code=end

File: leetcode/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
#include <bits/stdc++.h>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0;
        int l = 0;
        unordered_map<char, int> cnt(26);
        for (int i = 0; i < s.size(); i++) {
            cnt[s[i]]++;
            while (cnt[s[i]] > 1) {
                cnt[s[l++]]--;
            }
            ans = max(ans, i-l+1);
        }
        return ans;
    }
};
// @lc code=end

